tmpl/wiki/tutorial-lwt.md

```diff
@@ -74,508 +111,327 @@ Assuming you are building for POSIX using native kernel sockets, compile the app
+(* Handle a frame of data from the network... *)
+Lwt.async (fun () ->
+  Lwt.catch (fun () -> fn data)
+    (fun ex ->
+       Log.err (fun f -> f "uncaught exception from listen callback \
+                            while handling frame:@\n%a@\nException: @[%s@]"
+                   S.pp_frame data (Printexc.to_string ex));
+       Lwt.return ()
+    )
+)
 ```
 
-###Challenge
+By the way, the reason `async` and `catch` take functions rather than plain threads is so they can catch OCaml exceptions (not just failed threads). Otherwise, if we had `Lwt.catch (fn data) ...` and `fn data` used `raise` then the exception would occur before `Lwt.catch` ran, and it wouldn't be able to handle it.
 
-This `timeout` function does not allow one to use the result returned by the thread `t`.
 
-Modify the `timeout` function so that it returns either `None` if `t` has not yet returned after `f` seconds or `Some v` if `t` returns `v` within `f` seconds. In order to achieve this behaviour it is possible to use the function `state` that, given a thread, returns the state it is in, either `Sleep`, `Return` or `Fail`.
+##Error handling
 
-You can test your solution with this application, which creates a thread and that is cancelled before it returns:
+In Mirage code, we typically distinguish two types of error: programming errors (bugs, which should be reported to the programmer to be fixed) and expected errors (e.g. network disconnected or invalid TCP packet received).
+We try to use the type system to ensure that expected errors are handled gracefully.
 
-```
-  let t =  Time.sleep (Random.float 3.0) >> return "Heads" in
-    timeout 2.0 t >>= fun v ->
-    C.log c (match v with None -> "cancelled" | Some v -> v);
-    C.log c "Finished";
-    return ()
-```
+###Use result for expected errors
 
-###Solution
+For expected errors, you should use the `result` type, which provides `Ok` and `Error` constructors.
+This is a built-in in OCaml 4.03 and available from the `result` opam package for older versions.
 
-```
-  let timeout f t =
-    Time.sleep f >>
-    match state t with
-    | Return v -> return (Some v)
-    | _        -> cancel t; return None
+Here's an example that calls `read_arg` twice and returns the sum of the results on success. If either `read_arg` returns an error then that is returned immediately.
 
 ```
+let example () =
+  read_arg () >>= function
+  | Error _ as e -> Lwt.return e
+  | Ok a ->
+  read_arg () >>= function
+  | Error _ as e -> Lwt.return e
+  | Ok b ->
+  return (Ok (a + b))
+```
 
-This solution and application are found in `lwt/src/unikernels.ml` in the repository. Build with target `timeout1`.
-
-Does your solution match the one given here and always returns after `f` seconds, even when `t` returns within `f` seconds?
-
-This is a good place to introduce a third operation to compose threads: `pick`.
+It is often useful to provide some helpers to handle this pattern (using Lwt threads and result types together) more simply:
```

Comment:
Relevant: https://github.com/ocsigen/lwt/pull/247. Nothing to do about it here yet, though :)